src/MSBuild/MSBuild.csproj

```
@@ -330,4 +330,14 @@
 
   </Target>
 
+  <UsingTask TaskName="ValidateMSBuildPackageDependencyVersions" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
+    <Task>
+      <Code Source="ValidateMSBuildPackageDependencyVersions.cs" Language="cs" />
+    </Task>
+  </UsingTask>
+
+  <Target Name="ValidateMSBuildPackageDependencyVersions" BeforeTargets="AfterBuild" Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'" >
```

```suggestion
  <Target Name="ValidateMSBuildPackageDependencyVersions" AfterTargets="PrepareForBuild" Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'" >
```

This will get it to check the checked-in version, which is what we should be editing, rather than the copy with possibly added auto binding redirects (which we want to avoid).

---

So it seems like it also hasn't copied the packages to the output directory yet at that point. The two options I see:
Revert back to how it was before. Not ideal but easy.
Have two tasks: one that parses the app.config into an item and another that looks at the item and compares it to packages. This fits your ideal but isn't nearly as clean.
Thoughts?